options.go

```diff
@@ -17,77 +17,420 @@
+		usingCredentials = true
+
+		u, err := usr.Current()
+		if err != nil {
+			return nil, dbParameters{}, err
+		}
+
+		file := path.Join(u.HomeDir, ".edgedb", "credentials", dsn+".json")
+		data, err := ioutil.ReadFile(file)
+		if err != nil {
+			return nil, dbParameters{}, fmt.Errorf(
+				"cannot read credentials of instance %q: %v%w",
+				dsn, err, ErrClientMistake)
+		}
+
+		var creds credentials
+		if e := json.Unmarshal(data, &creds); e != nil {
+			return nil, dbParameters{}, fmt.Errorf(
+				"cannot read credentials of instance %q: %v%w",
+				dsn, e, ErrClientMistake)
+		}
+
+		if len(ports) == 0 {
+			ports = []int{creds.Port}
+		}
+
+		if user == "" {
+			user = creds.User
+		}
+
+		if len(hosts) == 0 && creds.Host != "" {
+			hosts = []string{creds.Host}
+		}
+
+		if password == "" {
+			password = creds.Password
+		}
+
+		if database == "" {
+			database = creds.Database
+		}
 	}
 
-	var port int
-	if parsed.Port() == "" {
-		port = 5656
-	} else {
-		port, err = strconv.Atoi(parsed.Port())
+	var err error
+
+	if spec := os.Getenv("EDGEDB_HOST"); len(hosts) == 0 && spec != "" {
+		hosts, ports, err = parseHostList(spec, ports)
 		if err != nil {
-			return opts, err
+			return nil, dbParameters{}, err
 		}
 	}
 
-	host := strings.Split(parsed.Host, ":")[0]
-	db := strings.TrimLeft(parsed.Path, "/")
-	password, _ := parsed.User.Password()
+	if len(hosts) == 0 {
+		if !usingCredentials {
+			hosts = append(hosts, defaultHosts...)
+		}
+		if !opts.admin {
+			hosts = append(hosts, "localhost")
+		}
+	}
 
-	return Options{
-		Host:     host,
-		Port:     port,
-		User:     parsed.User.Username(),
-		Database: db,
-		Password: password,
-	}, nil
+	if len(ports) == 0 {
+		if portSpec := os.Getenv("EDGEDB_PORT"); portSpec != "" {
+			ports, err = parsePortSpec(portSpec)
+			if err != nil {
+				return nil, dbParameters{}, err
+			}
+		} else {
+			ports = []int{5656}
+		}
+	}
+
+	ports, err = validatePortSpec(hosts, ports)
+	if err != nil {
+		return nil, dbParameters{}, err
+	}
+
+	if val := os.Getenv("EDGEDB_USER"); user == "" {
+		user = val
+	}
+
+	if val := os.Getenv("EDGEDB_PASSWORD"); password == "" {
+		password = val
+	}
+
+	if val := os.Getenv("EDGEDB_DATABASE"); database == "" {
+		database = val
+	}
+
+	if database == "" {
+		database = user
+	}
+
+	if user == "" {
+		return nil, dbParameters{}, fmt.Errorf(
+			"could not determine username to connect with%w", ErrBadConfig)
+	}
+
+	if database == "" {
+		return nil, dbParameters{}, fmt.Errorf(
+			"could not determine database to connect with%w", ErrBadConfig)
+	}
+
+	haveUnixSockets := false
+	var addrs []netParameters
+	for i := 0; i < len(hosts); i++ {
+		h := hosts[i]
+		p := ports[i]
+
+		if strings.HasPrefix(h, "/") {
+			if !strings.Contains(h, ".s.EDGEDB.") {
+				var sockName string
+				if opts.admin {
+					sockName = fmt.Sprintf(".s.EDGEDB.admin.%v", p)
```

Comment:
I'm not sure if we should support admin sockets in Go driver. @elprans ?

Reply:
Admin sockets are deprecated across the board, please don't add support.

Reply:
The admin option is not part of the public API. It is only there right now for running tests in CI.

Reply:
> The admin option is not part of the public API. It is only there right now for running tests in CI.

I suggest you to take a look how edgedb-js CI is set up. There's no need for admin sockets that way (at least it shouldn't be).

I'd prefer not adding them in any API, public or private.

Reply:
I tried removing the admin option and then figuring out how to make the tests connect to a server created with the `edgedb-server` command. I can't figure out what the default password is for either the `edgedb` user or the linux user. The edgedb-python test suite seems to be using the admin socket to set a password for the `edgedb` user.

https://github.com/edgedb/edgedb-python/blob/289592fb540d2f4257956948823a309596bbe058/edgedb/_testbase.py#L70
https://github.com/edgedb/edgedb-python/blob/289592fb540d2f4257956948823a309596bbe058/edgedb/_cluster.py#L242
https://github.com/edgedb/edgedb-python/blob/289592fb540d2f4257956948823a309596bbe058/edgedb/_cluster.py#L228

Is there something I am missing?

Reply:
I am running edgedb-server like this
```bash
edgedb-server-1-alpha6 -D /tmp/edgedb-go-test-server-006711182 --testmode --runstate-dir /tmp/edgedb-go-test-server-006711182 --port 49697
```

Reply:
hmm... @elprans ?

Reply:
You also want 

```
--bootstrap-script='CONFIGURE SYSTEM INSERT Auth {priority := 0, method := (INSERT Trust)}`
```

Reply:
**Or** (better):

```
--bootstrap-script='CREATE SUPERUSER ROLE test { SET password := "test-password" }'
```

Reply:
@elprans it doesn't look like `--bootstrap-script` is an option :confused: 

```bash
$ edgedb-server-1-alpha6 --temp-dir --testmode --echo-runtime-info --port=auto --auto-shutdown --bootstrap-script='CREATE SUPERUSER ROLE test { SET password := "shhh"  }'
Usage: edgedb-server-1-alpha6 [OPTIONS]
Try 'edgedb-server-1-alpha6 -h' for help.

Error: no such option: --bootstrap-script
```

```bash
$ edgedb-server-1-alpha6 -h
Usage: edgedb-server-1-alpha6 [OPTIONS]

Options:
  -D, --data-dir PATH             database cluster directory
  --postgres-dsn TEXT             DSN of a remote Postgres cluster, if using
                                  one

  -l, --log-level TEXT            Logging level.  Possible values: (d)ebug,
                                  (i)nfo, (w)arn, (e)rror, (s)ilent

  --log-to DEST                   send logs to DEST, where DEST can be a file
                                  name, "syslog", or "stderr"

  --bootstrap                     bootstrap the database cluster and exit
  --default-database TEXT         the name of the default database to create
  --default-database-user TEXT    the name of the default database owner
  --devmode / --no-devmode        enable or disable the development mode
  --testmode / --no-testmode      enable or disable the test mode
  -I, --bind-address TEXT         IP address to listen on
  -P, --port PORT                 port to listen on
  -b, --background                daemonize
  --pidfile-dir PATH              path to PID file directory
  --daemon-user INTEGER
  --daemon-group INTEGER
  --runstate-dir PATH             directory where UNIX sockets and other
                                  temporary runtime files will be placed
                                  (/run/edgedb by default)

  --max-backend-connections INTEGER
  --echo-runtime-info             echo runtime info to stdout; the format is
                                  JSON, prefixed by "EDGEDB_SERVER_DATA:",
                                  ended with a new line

  --temp-dir                      create a temporary database cluster
                                  directory that will be automatically purged
                                  on server shutdown

  --auto-shutdown                 shutdown the server after the last
                                  management connection is closed

  --max-protocol TEXT             maximum supported and advertized client
                                  protocol version

  --version                       Show the version and exit.
  -h, --help                      Show this message and exit.
```

Reply:
Oh, that's in nightly only, and it's actually `--bootstrap-command` (there's also `--bootstrap-script` taking a file path).

Reply:
I confirmed that I can make this work on a nightly build. Is everyone comfortable with the CI for this repo running against a nightly build? tentatively `edgedb-server-1-alpha6-dev5104`

Reply:
>  Is everyone comfortable with the CI for this repo running against a nightly build

Yes

> `edgedb-server-1-alpha6-dev5104`

This seems old.  `edgedb server install --nightly` reports `1.0a7.dev5243+g096c48c9-2020120300~focal.nightly`

Reply:
ok, using `edgedb-server-1-alpha7-dev5249`